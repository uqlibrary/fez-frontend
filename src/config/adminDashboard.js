import React from 'react';
import { default as adminDashboardLocale } from 'locale/components';

import Chip from '@mui/material/Chip';

import Today from 'modules/AdminDashboard/tabs/Today';
import SystemAlerts from 'modules/AdminDashboard/tabs/SystemAlerts';
import Reports from 'modules/AdminDashboard/tabs/Reports';

export const tabConfig = [
    { id: 0, title: adminDashboardLocale.components.adminDashboard.tabs.today.tabLabel, component: Today },
    {
        id: 1,
        title: adminDashboardLocale.components.adminDashboard.tabs.systemalerts.tabLabel,
        component: SystemAlerts,
    },
    { id: 2, title: adminDashboardLocale.components.adminDashboard.tabs.reports.tabLabel, component: Reports },
];

export const tabProps = [
    {
        id: 1,
        render: data => {
            const count = data?.systemalerts?.mine;
            return count
                ? {
                      icon: <Chip color="error" label={count} size="small" data-testid={`tab-counter-${count}`} />,
                      iconPosition: 'end',
                  }
                : {};
        },
    },
];
