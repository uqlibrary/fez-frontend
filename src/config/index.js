export {
    API_URL,
    APP_URL,
    AUTH_URL_LOGIN,
    AUTH_URL_LOGOUT,
    SESSION_COOKIE_NAME,
    SESSION_USER_GROUP_COOKIE_NAME,
    TOKEN_NAME,
    ORCID_BASE_URL,
    ORCID_CLIENT_ID,
    ORCID_AUTHORIZATION_URL,
    GOOGLE_MAPS_API_URL,
    GOOGLE_MAPS_API_CHINA_URL,
    PATH_PREFIX,
    DATASET_ACCESS_CONDITIONS_OPTIONS,
} from './general';

export { api, generateCancelToken, sessionApi } from './axios';
export * as validation from './validation';
export * as general from './general';
export { publicationTypes, contentIndicators, numberToWords } from './general';
export * as routes from './routes';
export * as openAccessConfig from './openAccess';
export { default as incompleteRecord } from './incompleteRecord';
export { viewRecordsConfig } from './viewRecord';
export { trendingPublicationsConfig } from './trendingPublications';
export { claimRecordConfig } from './claimRecord';
export { mui1theme, adminTheme } from './theme';
export * as BULK_UPDATES from './bulkUpdates';
export { pathConfig } from './pathConfig';
export { viewJournalConfig } from './viewJournal';
export { TRANSITION_COHORT } from './thesisSubmissionTransition';

export { default as imageGalleryConfig } from './imageGalleryConfig';
export { communityCollectionsConfig } from './communityCollections';
export { controlledVocabConfig } from './controlledVocabConfig';
