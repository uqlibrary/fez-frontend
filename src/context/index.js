import { useContext } from 'react';

import AccountContext from './AccountContext';
import EditableContext from './EditableContext';
import FormErrorsContext from './FormErrorsContext';
import FormValuesContext from './FormValuesContext';
import LocallyStoredReducerContext from './LocallyStoredReducerContext';
import OrcidSyncContext from './OrcidSyncContext';
import RecordContext from './RecordContext';
import RecordsSelectorContext from './RecordsSelectorContext';
import TabbedContext from './TabbedContext';
import ScrollToSectionContext from './ScrollToSectionContext';
import JournalContext from './JournalContext';

export const useAccountContext = () => useContext(AccountContext);
export const useEditableContext = () => useContext(EditableContext);
export const useFormErrorsContext = () => useContext(FormErrorsContext);
export const useFormValuesContext = () => useContext(FormValuesContext);
export const useLocallyStoredReducerContext = () => useContext(LocallyStoredReducerContext);
export const useOrcidSyncContext = () => useContext(OrcidSyncContext);
export const useRecordContext = () => useContext(RecordContext);
export const useRecordsSelectorContext = () => useContext(RecordsSelectorContext);
export const useTabbedContext = () => useContext(TabbedContext);
export const useScrollToSectionContext = () => useContext(ScrollToSectionContext);
export const useJournalContext = () => useContext(JournalContext);

export {
    AccountContext,
    EditableContext,
    FormErrorsContext,
    FormValuesContext,
    LocallyStoredReducerContext,
    OrcidSyncContext,
    RecordContext,
    RecordsSelectorContext,
    ScrollToSectionContext,
    TabbedContext,
    JournalContext,
};
