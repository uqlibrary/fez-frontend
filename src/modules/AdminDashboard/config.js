import React from 'react';
import moment from 'moment';

import Chip from '@mui/material/Chip';
import Typography from '@mui/material/Typography';

import { ExternalLink } from 'modules/SharedComponents/ExternalLink';

export const COLOURS = { assigned: '#338CFA', unassigned: '#B60DCE' };

export const LINK_UNPROCESSED_WORKS =
    'https://espace.library.uq.edu.au/records/search?page=1&pageSize=20&sortBy=score&sortDirection=Desc&searchQueryParams%5Ball%5D%5Bvalue%5D=ismemberof%3A+UQ%3A218198&searchQueryParams%5Ball%5D%5Blabel%5D=&searchMode=advanced';

export const animationTemplate = (i, duration, delay) =>
    `animateFadeIn ${duration}ms ease-out ${delay * (i + 1)}ms forwards`;

export const MENUACTIONS = {
    EDIT: 'EDIT',
    DELETE: 'DELETE',
    UP: 'MOVEUP',
    TOP: 'MOVETOP',
    DOWN: 'MOVEDOWN',
    BOTTOM: 'MOVEBOTTOM',
};

export const VIEWMODES = {
    VIEW: 'VIEW',
    ADD: 'ADD',
    EDIT: 'EDIT',
    DELETE: 'DELETE',
};
export const VIEWADMINPANELMODES = [VIEWMODES.ADD, VIEWMODES.EDIT, VIEWMODES.DELETE];
export const REORDERING = [MENUACTIONS.TOP, MENUACTIONS.UP, MENUACTIONS.BOTTOM, MENUACTIONS.DOWN];

export const DEFAULT_DATEPICKER_INPUT_FORMAT = 'DD/MM/YYYY';
export const DEFAULT_DATE_FORMAT = 'Do MMMM YYYY';
export const DEFAULT_DATE_FORMAT_WITH_TIME = 'Do MMMM YYYY hh:mm';
export const DEFAULT_DATE_FORMAT_WITH_TIME_24H = 'Do MMMM YYYY HH:mm';
export const DEFAULT_DATE_FORMAT_WITH_TIME_24H_SECONDS = 'Do MMMM YYYY HH:mm:ss';
export const DEFAULT_SERVER_DATE_FORMAT = 'YYYY-MM-DD HH:mm:ss';

export const SYSTEM_ALERT_ACTION = {
    ASSIGN: 'ASSIGN',
    RESOLVE: 'RESOLVE',
};

export const REPORT_TYPE = {
    systemalertlog: 1,
    workshistory: 2,
    workiddupe: 3,
    scopusiddupe: 4,
    doidupe: 5,
};

export const DEFAULT_SORTING = {
    alerts: [{ field: 'sat_created_date', sort: 'asc' }],
    systemalertlog: [{ field: 'sat_created_date', sort: 'asc' }],
    workshistory: [{ field: 'pre_date', sort: 'asc' }],
};

export const isUrl = str => {
    try {
        const newUrl = new URL(str);
        return newUrl.protocol === 'http:' || newUrl.protocol === 'https:';
    } catch (err) {
        return false;
    }
};

export const defaultLegacyReportOption = { sel_id: 0, sel_title: '', sel_description: '' };

export const optionDoubleRowRender = (props, option) => {
    return (
        <li
            {...props}
            style={{
                flexDirection: 'column',
                justifyContent: 'left',
                alignItems: 'flex-start',
                fontWeight: 400,
            }}
            data-testid={props.id}
        >
            <Typography variant="body1" color="textPrimary">
                {option.sel_title}
            </Typography>
            <Typography variant="body1" color="textSecondary">
                {option.sel_description}
            </Typography>
        </li>
    );
};

export const getReportTypeFromValue = value => Object.entries(REPORT_TYPE).find(arr => arr[1] === value)?.[0];
export const getDefaultSorting = reportType => DEFAULT_SORTING?.[reportType] || [];

export const getFormattedServerDate = (dateStr, format = DEFAULT_DATE_FORMAT) =>
    (dateStr && moment(dateStr, DEFAULT_SERVER_DATE_FORMAT).format(format)) || '';

export const getSystemAlertColumns = (locale, users) => {
    const alertStatus = locale.alertStatus;
    const alertStatusOption = Object.values(alertStatus);
    return [
        {
            field: 'sat_created_date',
            headerName: locale.columns.createdDate,
            width: 200,
            type: 'date',
            valueGetter: value => moment(value, DEFAULT_SERVER_DATE_FORMAT).toDate(),
            valueFormatter: value => moment(value).format(DEFAULT_DATE_FORMAT_WITH_TIME_24H),
        },
        { field: 'sat_title', headerName: locale.columns.topic, flex: 1 },
        {
            field: 'status',
            headerName: locale.columns.status,
            width: 160,
            valueGetter: (_, row) =>
                !!row.sat_assigned_to
                    ? users.find(user => user.id === row.sat_assigned_to)?.preferred_name ?? alertStatus.UNKNOWN
                    : alertStatus.UNASSIGNED,
            renderCell: params => (
                <Chip
                    data-testid={`alert-status-${params.id}`}
                    label={params.value}
                    variant="filled"
                    size="small"
                    color={alertStatusOption.includes(params.value) ? 'default' : 'primary'}
                />
            ),
        },
    ];
};

export const getDisplayReportColumns = ({ locale, actionState, params }) => {
    const report = actionState?.displayReport?.value || getReportTypeFromValue(params.report_type);

    const txt = locale.columns[report];
    switch (report) {
        case 'workshistory':
            return [
                { field: 'pre_id', headerName: txt.id, width: 150, exportOnly: true, exportOrder: 0 },
                { field: 'pre_pid', headerName: txt.pid, width: 150, order: 0, exportOrder: 1 },
                {
                    field: 'pre_date',
                    headerName: txt.dateCreated,
                    width: 150,
                    type: 'date',
                    valueGetter: value => (!!value && moment(value, DEFAULT_SERVER_DATE_FORMAT).toDate()) || null,
                    valueFormatter: value => {
                        if (value === null || value === '') return '';
                        return moment(value).format(DEFAULT_DATE_FORMAT);
                    },
                    order: 1,
                    exportOrder: 5,
                },
                {
                    field: 'rek_date',
                    headerName: txt.pubDate,
                    width: 150,
                    type: 'date',
                    valueGetter: value =>
                        (!!value && moment(value, DEFAULT_SERVER_DATE_FORMAT).format(DEFAULT_DATE_FORMAT)) || '',
                    exportOnly: true,
                    exportOrder: 4,
                },
                { field: 'rek_genre', headerName: txt.genre, minWidth: 300, flex: 1, order: 2, exportOrder: 2 },
                { field: 'rek_subtype', headerName: txt.subtype, minWidth: 300, flex: 1, order: 3, exportOrder: 3 },
                { field: 'usr_username', headerName: txt.user, width: 150, order: 4, exportOrder: 6 },
                { field: 'pre_detail', headerName: txt.action, minWidth: 600, flex: 1, order: 5, exportOrder: 7 },
            ];
        default:
            const systemIdParam = actionState?.systemAlertId || params?.record_id || '';
            const cols = [
                { field: 'sat_id', headerName: txt.id, order: 0, exportOrder: 0 },
                {
                    field: 'sat_created_date',
                    headerName: txt.dateCreated,
                    width: 150,
                    type: 'date',
                    valueGetter: value => (!!value && moment(value, DEFAULT_SERVER_DATE_FORMAT).toDate()) || null,
                    valueFormatter: value => {
                        if (value === null || value === '') return '';
                        return moment(value).format(DEFAULT_DATE_FORMAT);
                    },
                    order: 1,
                    exportOrder: 1,
                },
                {
                    field: 'assigned_to_username',
                    headerName: 'Assigned username',
                    width: 150,
                    exportOrder: 2,
                    exportOnly: true,
                },
                {
                    field: 'assigned_to_full_name',
                    headerName: txt.assignedTo,
                    width: 150,
                    exportOrder: 3,
                    order: 2,
                },
                {
                    field: 'sat_assigned_date',
                    headerName: txt.assignedDate,
                    width: 150,
                    type: 'date',
                    valueGetter: value => (!!value && moment(value, DEFAULT_SERVER_DATE_FORMAT).toDate()) || null,
                    valueFormatter: value => {
                        if (value === null || value === '') return '';
                        return moment(value).format(DEFAULT_DATE_FORMAT);
                    },
                    order: 3,
                    exportOrder: 4,
                },
                {
                    field: 'resolved_by_username',
                    headerName: 'Resolved username',
                    width: 150,
                    exportOrder: 5,
                    exportOnly: true,
                },
                {
                    field: 'resolved_by_full_name',
                    headerName: txt.resolvedBy,
                    width: 150,
                    order: 4,
                    exportOrder: 6,
                },
                {
                    field: 'sat_resolved_date',
                    headerName: txt.resolvedDate,
                    width: 150,
                    type: 'date',
                    valueGetter: value => (!!value && moment(value, DEFAULT_SERVER_DATE_FORMAT).toDate()) || null,
                    valueFormatter: value => {
                        if (value === null || value === '') return '';
                        return moment(value).format(DEFAULT_DATE_FORMAT);
                    },
                    order: 5,
                    exportOrder: 7,
                },
                { field: 'sat_title', headerName: txt.title, minWidth: 400, flex: 1, order: 6, exportOrder: 8 },
                {
                    field: 'sat_link',
                    headerName: txt.link,
                    minWidth: 400,
                    flex: 1,
                    renderCell: row =>
                        isUrl(row.value) ? (
                            <ExternalLink id={`link_${row.id}`} href={row.value} inline>
                                {row.value}
                            </ExternalLink>
                        ) : (
                            row.value
                        ),
                    order: 8,
                    exportOrder: 9,
                },
            ];

            if (systemIdParam !== '') {
                cols.push({
                    field: 'sat_content',
                    headerName: txt.content,
                    minWidth: 1000,
                    flex: 1,
                    order: 7,
                    exportOrder: 10,
                });
            }
            return cols;
    }
};
