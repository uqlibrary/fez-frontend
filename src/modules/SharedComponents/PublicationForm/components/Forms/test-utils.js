import { useForm } from 'react-hook-form';
import { render, WithReduxStore } from 'test-utils';
import React from 'react';
import { formValues } from '../PublicationForm';

export const setupWrapper = (Form, props) => {
    const Wrapper = () => {
        const {
            control,
            formState: { isSubmitting },
        } = useForm();

        // eslint-disable-next-line react/prop-types
        return <Form {...{ control, isSubmitting, ...props, formValues: formValues(props?.formValues || {}) }} />;
    };

    return render(
        <WithReduxStore>
            <Wrapper />
        </WithReduxStore>,
    );
};
