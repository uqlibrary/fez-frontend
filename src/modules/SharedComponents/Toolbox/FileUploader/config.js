export const FILE_ACCESS_CONDITION_CLOSED = 1;
export const FILE_ACCESS_CONDITION_OPEN = 5;
export const FILE_ACCESS_CONDITION_INHERIT = 99;
export const FILE_SECURITY_POLICY_ADMIN = 1;
export const FILE_SECURITY_POLICY_EVIDENCE_ASSESSORS = 3;
export const FILE_SECURITY_POLICY_PUBLIC = 5;

export const FILE_ACCESS_OPTIONS = [
    { text: 'Open Access', value: FILE_ACCESS_CONDITION_OPEN },
    { text: 'Closed Access', value: FILE_ACCESS_CONDITION_CLOSED },
];

export const INHERIT_OPTION = {
    text: 'Inherit from parent collection',
    value: FILE_ACCESS_CONDITION_INHERIT,
};

export const SIZE_BASE = 1000;

export const DEFAULT_FILE_UPLOAD_LIMIT = 10;
export const DEFAULT_MAX_FILE_SIZE = 5;

export const SIZE_UNIT_B = 'B';
export const SIZE_UNIT_K = 'K';
export const SIZE_UNIT_M = 'M';
export const SIZE_UNIT_G = 'G';

export const FILE_META_KEY_ACCESS_CONDITION = 'access_condition_id';
export const FILE_META_KEY_EMBARGO_DATE = 'date';
export const FILE_META_KEY_DESCRIPTION = 'description';
export const FILE_META_KEY_SECURITY_POLICY = 'security_policy';

export const SIZE_UNITS = [SIZE_UNIT_B, SIZE_UNIT_K, SIZE_UNIT_M, SIZE_UNIT_G];

export const FILE_NAME_RESTRICTION = /^(?=^\S*$)(?=^[a-z\d\-_]+(\.\d{3}|\.r\d{2,3}|\.part\d{1,3})?\.[^\.]+$)(?=.{1,45}$)(?!(web_|preview_|thumbnail_|stream_|fezacml_|presmd_|\d))[a-z\d\-_\.]+/;

/**
 * When updating the following list, please make sure to update it as well in the following projects:
 * - https://github.com/uqlibrary/fez-frontend
 * - https://github.com/uqlibrary/lambda-fez-upload-handler-trigger
 * - https://github.com/uqlibrary/api
 * - https://github.com/uqlibrary/fez
 *
 * Note: ext -> mimetype is a 1-n relation, but we only care about one value
 */
export const MIME_TYPE_WHITELIST = {
    pdf: 'application/pdf',
    jpe: 'image/jpeg',
    jpeg: 'image/jpeg',
    jpg: 'image/jpeg',
    gif: 'image/gif',
    tif: 'image/tiff',
    tiff: 'image/tiff',
    png: 'image/png',
    zip: 'application/zip',
    tar: 'application/x-tar',
    gz: 'application/gzip',
    '7z': 'application/x-7z-compressed',
    rar: 'application/x-rar-compressed',
    mk3d: 'video/x-matroska',
    mks: 'video/x-matroska',
    mkv: 'video/x-matroska',
    qt: 'video/quicktime',
    mov: 'video/quicktime',
    mp4: 'video/mp4',
    mp4v: 'video/mp4',
    mp3: 'audio/mp3',
    mxf: 'application/mxf',
    wav: 'audio/wav',
    mpe: 'video/mpeg',
    mpeg: 'video/mpeg',
    mpg: 'video/mpeg',
    m1v: 'video/mpeg',
    m2v: 'video/mpeg',
    wmv: 'video/x-ms-wmv',
    wma: 'audio/x-ms-wma',
    avi: 'video/x-msvideo',
    m4a: 'audio/x-m4a',
    csv: 'text/csv',
    xla: 'application/vnd.ms-excel',
    xlc: 'application/vnd.ms-excel',
    xlm: 'application/vnd.ms-excel',
    xls: 'application/vnd.ms-excel',
    xlt: 'application/vnd.ms-excel',
    xlw: 'application/vnd.ms-excel',
    xlsx: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
    ods: 'application/vnd.oasis.opendocument.spreadsheet',
    gsheet: 'application/vnd.google-apps.spreadsheet',
};
