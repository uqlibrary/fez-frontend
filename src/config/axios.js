import axios from 'axios';
import Cookies from 'js-cookie';
import { setupCache } from 'axios-cache-interceptor';
import { API_URL, SESSION_COOKIE_NAME, SESSION_USER_GROUP_COOKIE_NAME, TOKEN_NAME } from './general';
import { store } from 'config/store';
import { logout } from 'actions/account';
import { showAppAlert } from 'actions/app';
import locale from 'locale/global';
import * as Sentry from '@sentry/react';
import param from 'can-param';
import { pathConfig } from 'config/pathConfig';
import { isDevEnv, isTest } from '../helpers/general';
import { FIELD_OF_RESEARCH_VOCAB_ID, AIATSIS_CODES_VOCAB_ID } from 'config/general';

let apiClient = axios.create({
    baseURL: API_URL,
    crossdomain: true,
});

if (!isDevEnv() && !isTest()) {
    // note: axios-cache-interceptor is not compatible with axios response mocks & tests
    // upon updating it or changing config settings, make sure to test it using prodtest env
    apiClient = setupCache(apiClient, {
        // the option below only works when importing from "axios-cache-interceptor.dev"
        // (make sure to disable stripping of console.* funcs in webpack-dist.config.json)
        // debug: dc,
        ttl: 15 * 60 * 1000,
    });

    // the place the below is declared matters - see https://axios-cache-interceptor.js.org/guide/interceptors
    const nonCachedRoutes = ['records/search', 'journals/search', 'orcid'];
    const ignoreServerHeaderRoutes = [`vocabularies?cvo_ids=${FIELD_OF_RESEARCH_VOCAB_ID},${AIATSIS_CODES_VOCAB_ID}`];
    apiClient.interceptors.request.use(request => {
        const queryStringParams = Object.keys(request.params || {});
        if (!!request.cache && ignoreServerHeaderRoutes.find(route => request.url.includes(route))) {
            request.cache.interpretHeader = false;
        } else if (
            !!request.cache &&
            // disabled it when querystring params are present or when it partially matches a non cached route
            (queryStringParams.length ||
                request.url.includes('?') ||
                nonCachedRoutes.find(route => request.url.includes(route)))
        ) {
            /* eslint-disable max-len */
            // dc(`disabling cache for: ${request.url}${queryStringParams.length ? `?${JSON.stringify(request.params)}` : ''}`);
            // disabled cache
            request.cache = false;
            return request;
        }
        /* eslint-disable max-len */
        // dc(`the following request will be cached: ${request.url}${queryStringParams.length ? `?${JSON.stringify(request.params)}` : ''}`);
        return request;
    });
}

export const api = apiClient;

export const sessionApi = axios.create({
    baseURL: API_URL,
    crossdomain: true,
});

// need to generate a new token for each request otherwise if you try a new request with the old token,
// axios will appear to cancel your request automatically
export const generateCancelToken = () => {
    const CancelToken = axios.CancelToken;
    return CancelToken.source();
};

export const setupDefaults = () => {
    // If there is a local cookie available, then set the api headers for x-uql-token
    if (!!Cookies.get(SESSION_COOKIE_NAME) && !!Cookies.get(SESSION_USER_GROUP_COOKIE_NAME)) {
        api.defaults.headers.common[TOKEN_NAME] = Cookies.get(SESSION_COOKIE_NAME);
        sessionApi.defaults.headers.common[TOKEN_NAME] = Cookies.get(SESSION_COOKIE_NAME);
    }
    // allow us to safely force a given SESSION_COOKIE_NAME during development
    if (process.env.NODE_ENV === 'development' && !!process.env.SESSION_COOKIE_NAME) {
        api.defaults.headers.common[TOKEN_NAME] = process.env.SESSION_COOKIE_NAME;
        sessionApi.defaults.headers.common[TOKEN_NAME] = process.env.SESSION_COOKIE_NAME;
    }
};
setupDefaults();

api.isCancel = axios.isCancel; // needed for cancelling requests and the instance created does not have this method

export let lastRequest = null;
let isGet = null;
api.interceptors.request.use(request => {
    lastRequest = request;
    isGet = request.method === 'get';
    if (
        (request.url?.includes('records/search') || request.url?.includes('records/export')) &&
        !!request.params &&
        !!request.params.mode &&
        request.params.mode === 'advanced'
    ) {
        request.paramsSerializer = { serialize: params => param(params) };
    }
    return request;
});

const reportToSentry = error => {
    let detailedError = '';
    if (error.response) {
        detailedError = `Data: ${JSON.stringify(error.response.data)}; Status: ${
            error.response.status
        }; Headers: ${JSON.stringify(error.response.headers)}`;
    } else {
        detailedError = `Something happened in setting up the request that triggered an Error: ${error.message}`;
    }

    Sentry.withScope(scope => {
        scope.setExtras({ error: detailedError });
        Sentry.captureException(error);
    });
};

/**
 * Return an instance of an alike in built-in Error class with a public message param and additional given properties
 *
 * @param message
 * @param extras
 * @return {{}}
 */
export const createSentryFriendlyError = (message, extras = {}) =>
    new (class {
        constructor(message) {
            Object.assign(this, { ...extras, message, stack: new Error(message).stack });
        }
    })(message);

api.interceptors.response.use(
    response => {
        if (!isGet) {
            const promise = Promise.resolve(response.status === 201 ? response : response.data);
            return api.store?.clear ? api.store?.clear().then(() => promise) : promise;
        }
        return Promise.resolve(response.data);
    },
    error => {
        // 403 for tool api lookup is handled in actions/thirdPartyLookupTool.js
        const handlesErrorsInternally =
            !error?.config ||
            (error.config.url && error.config.url.includes(pathConfig.admin.thirdPartyTools.slice(1)));

        const reportHttpStatusToSentry = [422];
        if (!!error?.response?.status && reportHttpStatusToSentry.includes(error.response.status)) {
            reportToSentry(error);
        }

        let errorMessage = null;
        const errorStatus = error?.response?.status || -1;
        if (!handlesErrorsInternally) {
            if (errorStatus === 401) {
                if (!!Cookies.get(SESSION_COOKIE_NAME)) {
                    Cookies.remove(SESSION_COOKIE_NAME, { path: '/', domain: '.library.uq.edu.au' });
                    Cookies.remove(SESSION_USER_GROUP_COOKIE_NAME, { path: '/', domain: '.library.uq.edu.au' });
                    delete api.defaults.headers.common[TOKEN_NAME];
                }

                if (process.env.NODE_ENV === 'test' || process.env.NODE_ENV === 'cc') {
                    global.mockActionsStore.dispatch(logout());
                } else {
                    store.dispatch(logout());
                }
            }

            if (!!error.message && errorStatus === 500) {
                errorMessage =
                    ((error.response || {}).data || {}).message || locale.global.errorMessages[error.response.status];
                if (process.env.NODE_ENV === 'test' || process.env.NODE_ENV === 'cc') {
                    global.mockActionsStore.dispatch(showAppAlert(error.response.data));
                } else {
                    store.dispatch(showAppAlert(error.response.data));
                }
            } else if (!!error.response && !!error.response.status) {
                const statusMessages = locale.global.errorMessages;
                errorMessage = statusMessages.hasOwnProperty(errorStatus)
                    ? statusMessages[errorStatus]
                    : statusMessages.generic;
                if ([410, 422].includes(errorStatus)) {
                    // override for these statuses
                    errorMessage = {
                        ...errorMessage,
                        ...error.response.data,
                    };
                }
            }
        }

        const shouldNotAppearInSentry =
            document.location.hostname === 'localhost' || // testing on AWS sometimes fires these
            [401, 403, 404, 410].includes(errorStatus) || // login expired - no notice required
            errorStatus === 0 || // catch those "the network request was interrupted" we see so much
            errorStatus === '0' || // don't know what format it comes in
            errorStatus === 500 || // api should handle these
            errorStatus === 502; // connection timed out - it happens, FE can't do anything about it
        if (!shouldNotAppearInSentry) {
            reportToSentry(error);
        }

        if (!!errorMessage) {
            return Promise.reject(
                createSentryFriendlyError(errorMessage?.message || null, {
                    request: error.request,
                    ...errorMessage,
                    // allow the original error message to be handled further down the stack
                    ...(error.response?.data ? { original: error.response.data } : {}),
                }),
            );
        } else {
            return Promise.reject(error);
        }
    },
);
