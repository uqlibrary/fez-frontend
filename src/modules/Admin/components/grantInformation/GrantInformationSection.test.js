import React from 'react';
import { rtlRender, FormProviderWrapper, assertDisabled, assertEnabled } from 'test-utils';
import GrantInformationSection from './GrantInformationSection';

jest.mock('../../../../context');
import { useRecordContext } from 'context';

import { PUBLICATION_TYPE_JOURNAL_ARTICLE } from 'config/general';

function setup(testProps = {}, renderer = rtlRender) {
    const props = {
        ...testProps,
    };

    return renderer(
        <FormProviderWrapper values={{}}>
            <GrantInformationSection {...props} />
        </FormProviderWrapper>,
    );
}

describe('GrantInformationSection', () => {
    beforeEach(() => {
        useRecordContext.mockImplementation(() => ({
            record: {
                rek_display_type: PUBLICATION_TYPE_JOURNAL_ARTICLE,
            },
        }));
    });

    afterEach(() => {
        useRecordContext.mockReset();
    });

    it('should render default view', () => {
        const { container, getByTestId } = setup({});
        assertEnabled(getByTestId('rek-grant-agency-input'));
        expect(container).toMatchSnapshot();
    });

    it('should render disabled view', () => {
        const { container, getByTestId } = setup({ disabled: true });
        assertDisabled(getByTestId('rek-grant-agency-input'));
        expect(container).toMatchSnapshot();
    });
});
