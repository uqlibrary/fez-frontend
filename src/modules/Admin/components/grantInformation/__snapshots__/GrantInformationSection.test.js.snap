// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrantInformationSection should render default view 1`] = `
<Section
  cards={
    [
      {
        "groups": [
          [
            "grants",
          ],
        ],
        "title": "Grant information",
      },
    ]
  }
  disabled={false}
/>
`;

exports[`GrantInformationSection should render disabled view 1`] = `
<Section
  cards={
    [
      {
        "groups": [
          [
            "grants",
          ],
        ],
        "title": "Grant information",
      },
    ]
  }
  disabled={true}
/>
`;
